Make eventsIfTopMost also work for manageHoverState, and introduce

dabreegster opened the pull request:
openIfTopMost for Popups.

This PR does two separate things:
1) Make the existing `eventsIfTopMost` option also apply to setting hover state
2) Introduce a new `openIfTopMost` for Popups that works the same way

The existing demo page is extended to show all of this:

https://github.com/dimfeld/svelte-maplibre/assets/1664407/2c2046cc-1034-45eb-bde9-42fedab0a67e

dabreegster commented on src/routes/examples/overlapping_layer_events/+page.svelte:
```diff
@@ -46,13 +48,30 @@
+  When layers overlap, only activate the popup for top-most layer
+</label>
+
+<fieldset class="border border-gray-300 self-start px-2 flex gap-x-4 mb-2">
```
This winds up aligned left and doesn't look great. I am awful at CSS, help much appreciated

stuartlynn replied:
I think for center align you want to add items-center ( I am assuming this is tailwind classes)

dabreegster commented on src/routes/examples/overlapping_layer_events/+page.svelte:
```diff
@@ -66,12 +85,14 @@
 >
   {#each layers as { data, color }, i}
-    <GeoJson id="layer{i + 1}" data={{ type: 'FeatureCollection', features: data }}>
+    <GeoJson id="layer{i + 1}" data={{ type: 'FeatureCollection', features: data }} generateId>
```
`generateId` necessary for hover state

dabreegster commented on src/routes/examples/overlapping_layer_events/+page.svelte:
```diff
@@ -82,7 +103,20 @@
+        <Popup {openOn} {openIfTopMost} let:features>
+          <div style:background={color} style:color="white">
+            <p>{features.length} features from {color} layer</p>
+            {#if color == 'red'}
```
To more clearly show overlapping popups, it's useful to make the lower ones larger. This was the simplest way I could think to do that

dimfeld replied:
Yeah this is tricky since you need to

1. Override Maplibre's popup positioning (could be done with offset I guess but the arrows would be wrong).
2. Add some layout engine to move them aside when multiple popups are visible
3. Do this only after the Popup has rendered, since before the first render you don't know how large it is
4. Make sure the popups aren't moved so far that they are off the screen

For these reasons, and since multiple popups at once is not really a common case anyway, I think it might actually be better to make `openIfTopMost` default to `true`. Any thoughts on that?

dabreegster replied:
To clarify, I don't think making multiple popups line up is a useful feature necessarily. I added coloring and sizing to the popups in this example just to demonstrate the bug/behavior when `openIfTopMost` is false.

Stuart also raised the idea that a useful UX pattern could be for one popup to receive and handle features from all layers. The real use case here is my main work project (which is now using svelte-maplibre!!), https://acteng.github.io/atip/browse.html. There are a bunch of overlapping layers that can be enabled, and currently the popup covers the top-most one. It might be less confusing and more useful (for someone using this app to understand small linear/point infrastructure or count data in context of a larger census-polygon area) to show info for everything under the cursor.

I think `openIfTopMost` is an intuitive default, so yes, I'll enable that now!

dabreegster replied:
Edit: I changed the changeset type to major, since this seems like a breaking change to popup behavior

dabreegster commented on src/lib/Popup.svelte:
```diff
@@ -21,6 +21,8 @@
   /** Define when to open the popup. If set to manual, you can open the popup programmatically by
    * setting the `open` attribute. */
   export let openOn: 'hover' | 'click' | 'dblclick' | 'contextmenu' | 'manual' = 'click';
+  /** Only open the popup if there's no feature from a higher layer covering this one. */
```
Oh yeah, another question: should this even be a separate option, or just inherit the layer's `eventsIfTopMost`?

stuartlynn replied:
I can imagine wanting different behaviours depending on the user action. For example if you still want to capture events from all layers to use in some other part of the app but only show the top level hover / click features.

dabreegster commented:
A bug I'm not sure how to fix: if one layer has `hoverCursor` and another doesn't, as you transition between overlapping features for the two, the cursor isn't un-set when it should be

dabreegster commented:
I rebased, adding in the `sourceLayer` fix to one spot.

This is imperfect with the `hoverCursor` problem, but I think it's a good enough start

dimfeld commented:
I would need to try it but I think adding 

```
      if (hoverCursor) {
        $map.getCanvas().style.cursor = hoverCursor;
      }
```

into the mousemove handler in `Layer.svelte` would fix the hoverCursor thing.

dimfeld commented:
I left [a comment above](https://github.com/dimfeld/svelte-maplibre/pull/71#discussion_r1345040120) about the default value for `openIfTopMost`. Curious to get your thoughts on that but after that I think this is good to go.

dabreegster commented:
I fixed the cursor behavior, and updated the example. The green layer has no cursor set, to demonstrate why it's necessary to reset to the empty string sometimes too. Now the cursor should always match the layer being hovered on.

Thanks for the help with the CSS alignment and tip about the cursor fix!

dimfeld commented:
Looks good. Thank you again!
